#* 
#* author: Tom Roeder
#* email: tmroeder@google.com
#* date: 2014-11-05T16:46:42-0800
#* 
#- Add support for binary Speaksfor in C++.
#- 
#- This commit also adds a test to the Go code that deserializes a
#- base64w-serialized output from the new encode_auth test app.
#- 
#- 
#- Add a C++ child app that runs under the Go Tao.
#- 
#- The go_child app can be run using host.sh in the environment set up by the Linux
#- host Tao from install.sh. It communicates with the Go Tao using fds 3 and 4,
#- like a normal Go-based Tao application.
#- 
#- 
#- Clean up C++ library to remove old code.
#- 
#- This commit removes all the C++ Tao implementation code, since the
#- implementation itself has moved to Go. The remaining code is meant to support
#- communication between a C++ program and the Go Tao, and no more than that.
#- 
#- 
#- Remove more superfluous code.
#- 
#- This commit cuts out more unused code from libtao.
#- 
#- 
#- Test attestation to the Go Tao.
#- 
#- This commit adds to the go_child C++ test a marshaled Speaksfor message using
#- the name of the Tao. It prints out the resulting attestation in base64w.
#- 
#- 
#- Remove unused tests and reformat files.
#- 
#- Applied clang-format to files.
#- 
#- 
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index a58cf88..b86b546 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -21,19 +21,6 @@ include_directories(${PROTOBUF_INCLUDE_DIRS})
 
 include(cloudproxy.cmake)
 
-# This macro assumes that the header_file will go in CMAKE_CURRENT_BINARY_DIR
-# and that the lua_file is located in CMAKE_CURRENT_SOURCE_DIR.
-macro(add_lua_header header_file lua_file)
-add_custom_command(
-    OUTPUT ${header_file}
-    COMMAND ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/bin2c -o ${CMAKE_CURRENT_BINARY_DIR}/${header_file} ${lua_file}
-    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
-    DEPENDS bin2c ${lua_file}
-   )
-endmacro()
-
-
 add_subdirectory(apps)
-add_subdirectory(cloudproxy)
 add_subdirectory(tao)
 add_subdirectory(third_party)
diff --git a/src/apps/CMakeLists.txt b/src/apps/CMakeLists.txt
index f2b831f..97d75be 100644
--- a/src/apps/CMakeLists.txt
+++ b/src/apps/CMakeLists.txt
@@ -21,27 +21,8 @@ include_directories(../third_party/gflags/src)
 add_definitions(-DOS_POSIX)
 include_directories(${CMAKE_SOURCE_DIR}/third_party/chromium/include)
 
-add_executable(tpm_tao tpm_tao.cc)
-target_link_libraries(tpm_tao tao)
+add_executable(encode_auth encode_auth.cc)
+target_link_libraries(encode_auth tao)
 
-add_executable(soft_tao soft_tao.cc)
-target_link_libraries(soft_tao tao)
-
-add_executable(get_pcrs get_pcrs.cc)
-target_link_libraries(get_pcrs tao)
-
-add_executable(linux_host linux_host.cc)
-target_link_libraries(linux_host tao)
-
-add_executable(demo demo.cc)
-target_link_libraries(demo tao)
-
-PROTOBUF_GENERATE_CPP(PROTO_SRCS PROTO_HDRS demo_server.proto)
-add_executable(demo_server demo_server.cc ${PROTO_SRCS} ${PROTO_HDRS})
-target_link_libraries(demo_server cloudproxy)
-
-add_executable(log_net_server log_net_server.cc)
-target_link_libraries(log_net_server tao)
-
-add_executable(tao_admin tao_admin.cc)
-target_link_libraries(tao_admin tao)
+add_executable(go_child go_child.cc)
+target_link_libraries(go_child tao)
diff --git a/src/apps/client.cc b/src/apps/client.cc
deleted file mode 100644
index 7fabfd3..0000000
diff --git a/src/apps/demo.cc b/src/apps/demo.cc
deleted file mode 100644
index df1c301..0000000
diff --git a/src/apps/demo_server.cc b/src/apps/demo_server.cc
deleted file mode 100644
index 7a047a9..0000000
diff --git a/src/apps/demo_server.proto b/src/apps/demo_server.proto
deleted file mode 100644
index 602e5d3..0000000
diff --git a/src/apps/encode_auth.cc b/src/apps/encode_auth.cc
new file mode 100644
index 0000000..b0351e2
--- /dev/null
+++ b/src/apps/encode_auth.cc
@@ -0,0 +1,52 @@
+//  Copyright (c) 2014, Google Inc.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include <iostream>
+
+#include <gflags/gflags.h>
+#include <glog/logging.h>
+
+#include "tao/util.h"
+
+using std::string;
+
+using tao::Base64WEncode;
+using tao::InitializeApp;
+using tao::MarshalKeyPrin;
+using tao::MarshalSpeaksfor;
+
+int main(int argc, char **argv) {
+  InitializeApp(&argc, &argv, false);
+
+  // A fake key for the parent.
+  string taoKeyName("test tao key");
+  string taoName;
+  if (!MarshalKeyPrin(taoKeyName, &taoName)) {
+    LOG(FATAL) << "Could not marshal a fake key auth.Prin value";
+  }
+
+  // A dummy key string to encode as bytes in MarshalSpeaksfor
+  string testKey("test key");
+  string speaksfor;
+  if (!MarshalSpeaksfor(testKey, taoName, &speaksfor)) {
+    LOG(FATAL) << "Could not marshal a speaksfor statement";
+  }
+
+  string encoded;
+  if (!Base64WEncode(speaksfor, &encoded)) {
+    LOG(FATAL) << "Could not encode the speaksfor in Base64W";
+  }
+
+  std::cout << encoded << std::endl;
+  return 0;
+}
diff --git a/src/apps/fclient.cc b/src/apps/fclient.cc
deleted file mode 100644
index 89d5a20..0000000
diff --git a/src/apps/fserver.cc b/src/apps/fserver.cc
deleted file mode 100644
index b531cd9..0000000
diff --git a/src/apps/get_guest_hash.cc b/src/apps/get_guest_hash.cc
deleted file mode 100644
index 1393949..0000000
diff --git a/src/apps/get_pcrs.cc b/src/apps/get_pcrs.cc
deleted file mode 100644
index 65fbd5d..0000000
diff --git a/src/apps/go_child.cc b/src/apps/go_child.cc
new file mode 100644
index 0000000..a92dab1
--- /dev/null
+++ b/src/apps/go_child.cc
@@ -0,0 +1,118 @@
+//  Copyright (c) 2014, Google Inc.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+#include <string>
+
+#include <gflags/gflags.h>
+#include <glog/logging.h>
+
+#include "tao/fd_message_channel.h"
+#include "tao/tao_rpc.h"
+#include "tao/util.h"
+
+using std::string;
+using std::unique_ptr;
+
+using tao::Base64WDecode;
+using tao::Base64WEncode;
+using tao::FDMessageChannel;
+using tao::InitializeApp;
+using tao::MarshalSpeaksfor;
+using tao::Tao;
+using tao::TaoRPC;
+
+int main(int argc, char **argv) {
+  InitializeApp(&argc, &argv, false);
+
+  // This code expects fd 3 and 4 to be the pipes from and to the Tao, so it
+  // doesn't need to take any parameters. It will establish a Tao Child Channel
+  // directly with these fds.
+  unique_ptr<FDMessageChannel> msg(new FDMessageChannel(3, 4));
+  unique_ptr<Tao> tao(new TaoRPC(msg.release()));
+  string bytes;
+  if (!tao->GetRandomBytes(10, &bytes)) {
+    LOG(FATAL) << "Couldn't get 10 bytes from the Tao RPC channel";
+  }
+
+  if (bytes.size() == 10) {
+    LOG(INFO) << "Got 10 bytes from the Tao RPC channel";
+  } else {
+    LOG(FATAL) << "Got " << bytes.size() << " bytes from the channel, but "
+                                            "expected 10";
+  }
+
+  string encodedBytes;
+  if (!Base64WEncode(bytes, &encodedBytes)) {
+    LOG(FATAL) << "Couldn't encode 10 bytes in Base64W";
+  }
+  LOG(INFO) << "Encoded bytes: " << encodedBytes;
+
+  string sealed;
+  if (!tao->Seal(bytes, Tao::SealPolicyDefault, &sealed)) {
+    LOG(FATAL) << "Couldn't seal bytes across the channel";
+  }
+
+  string encodedSealed;
+  if (!Base64WEncode(sealed, &encodedSealed)) {
+    LOG(FATAL) << "Couldn't encode the sealed bytes";
+  }
+  LOG(INFO) << "Encoded sealed bytes: " << encodedSealed;
+
+  string unsealed;
+  string policy;
+  if (!tao->Unseal(sealed, &unsealed, &policy)) {
+    LOG(FATAL) << "Couldn't unseal the tao-sealed data";
+  }
+  LOG(INFO) << "Got a seal policy '" << policy << "'";
+
+  if (policy.compare(Tao::SealPolicyDefault) != 0) {
+    LOG(FATAL) << "The policy returned by Unseal didn't match the Seal policy";
+  }
+
+  if (unsealed.compare(bytes) != 0) {
+    LOG(FATAL) << "The unsealed data didn't match the sealed data";
+  }
+
+  string encodedUnsealed;
+  if (!Base64WEncode(unsealed, &encodedUnsealed)) {
+    LOG(FATAL) << "Couldn't encoded the unsealed bytes";
+  }
+
+  LOG(INFO) << "Encoded unsealed bytes: " << encodedUnsealed;
+
+  // Set up a fake attestation using a fake key.
+  string taoName;
+  if (!tao->GetTaoName(&taoName)) {
+    LOG(FATAL) << "Couldn't get the name of the Tao";
+  }
+
+  string fakeKey("This is a fake key");
+  string msf;
+  if (!MarshalSpeaksfor(fakeKey, taoName, &msf)) {
+    LOG(FATAL) << "Couldn't marshal a speaksfor statement";
+  }
+
+  string attest;
+  if (!tao->Attest(msf, &attest)) {
+    LOG(FATAL) << "Couldn't attest to a fake key delegation";
+  }
+
+  string encodedAttest;
+  if (!Base64WEncode(attest, &encodedAttest)) {
+    LOG(FATAL) << "Couldn't encode the attestation";
+  }
+
+  LOG(INFO) << "Got attestation " << encodedAttest;
+
+  LOG(INFO) << "All Go Tao tests pass";
+  return 0;
+}
diff --git a/src/apps/http_echo_server.cc b/src/apps/http_echo_server.cc
deleted file mode 100644
index 5766353..0000000
diff --git a/src/apps/https_echo_server.cc b/src/apps/https_echo_server.cc
deleted file mode 100644
index 8aa4623..0000000
diff --git a/src/apps/linux_host.cc b/src/apps/linux_host.cc
deleted file mode 100644
index b067f49..0000000
diff --git a/src/apps/linux_kvm_guest_tao_service.cc b/src/apps/linux_kvm_guest_tao_service.cc
deleted file mode 100644
index 37fa3fe..0000000
diff --git a/src/apps/linux_kvm_tao_service.cc b/src/apps/linux_kvm_tao_service.cc
deleted file mode 100644
index 4ec4b9b..0000000
diff --git a/src/apps/log_net_server.cc b/src/apps/log_net_server.cc
deleted file mode 100644
index 78aafe9..0000000
diff --git a/src/apps/server.cc b/src/apps/server.cc
deleted file mode 100644
index c6b0e0c..0000000
diff --git a/src/apps/soft_tao.cc b/src/apps/soft_tao.cc
deleted file mode 100644
index 491ee7f..0000000
diff --git a/src/apps/tao_admin.cc b/src/apps/tao_admin.cc
deleted file mode 100644
index 00a91c7..0000000
diff --git a/src/apps/tcca.cc b/src/apps/tcca.cc
deleted file mode 100644
index d3acc3d..0000000
diff --git a/src/apps/tpm_tao.cc b/src/apps/tpm_tao.cc
deleted file mode 100644
index 837df32..0000000
diff --git a/src/cloudproxy/CMakeLists.txt b/src/cloudproxy/CMakeLists.txt
deleted file mode 100644
index d6dd681..0000000
diff --git a/src/cloudproxy/cloud_channel.cc b/src/cloudproxy/cloud_channel.cc
deleted file mode 100644
index dad0d5f..0000000
diff --git a/src/cloudproxy/cloud_channel.h b/src/cloudproxy/cloud_channel.h
deleted file mode 100644
index 6a8478e..0000000
diff --git a/src/cloudproxy/cloud_channel.proto b/src/cloudproxy/cloud_channel.proto
deleted file mode 100644
index 054d52e..0000000
diff --git a/src/cloudproxy/cloud_client.cc b/src/cloudproxy/cloud_client.cc
deleted file mode 100644
index c3a90ef..0000000
diff --git a/src/cloudproxy/cloud_client.h b/src/cloudproxy/cloud_client.h
deleted file mode 100644
index 7e7faa7..0000000
diff --git a/src/cloudproxy/cloud_client_server_unittests.cc b/src/cloudproxy/cloud_client_server_unittests.cc
deleted file mode 100644
index c1ac10a..0000000
diff --git a/src/cloudproxy/cloud_server.cc b/src/cloudproxy/cloud_server.cc
deleted file mode 100644
index 4ed0072..0000000
diff --git a/src/cloudproxy/cloud_server.h b/src/cloudproxy/cloud_server.h
deleted file mode 100644
index 2bc5eee..0000000
diff --git a/src/cloudproxy/cloudproxy_test.cc b/src/cloudproxy/cloudproxy_test.cc
deleted file mode 100644
index 2c65242..0000000
diff --git a/src/cloudproxy/tls_message_channel.cc b/src/cloudproxy/tls_message_channel.cc
deleted file mode 100644
index da232ec..0000000
diff --git a/src/cloudproxy/tls_message_channel.h b/src/cloudproxy/tls_message_channel.h
deleted file mode 100644
index 3a55a65..0000000
diff --git a/src/cloudproxy/util.cc b/src/cloudproxy/util.cc
deleted file mode 100644
index fb8d5e7..0000000
diff --git a/src/cloudproxy/util.h b/src/cloudproxy/util.h
deleted file mode 100644
index a48d7f6..0000000
diff --git a/src/cloudproxy/util_unittests.cc b/src/cloudproxy/util_unittests.cc
deleted file mode 100644
index 064ba76..0000000
diff --git a/src/cpcrypto/Test/DIRECTORYNOTICE.txt b/src/cpcrypto/Test/DIRECTORYNOTICE.txt
deleted file mode 100644
index d50133d..0000000
diff --git a/src/cpcrypto/Test/aesspeed.mak b/src/cpcrypto/Test/aesspeed.mak
deleted file mode 100755
index 22c2019..0000000
diff --git a/src/cpcrypto/Test/aesspeedtest.cc b/src/cpcrypto/Test/aesspeedtest.cc
deleted file mode 100755
index 953dac0..0000000
diff --git a/src/cpcrypto/Test/aestest.cc b/src/cpcrypto/Test/aestest.cc
deleted file mode 100755
index a8b08fc..0000000
diff --git a/src/cpcrypto/Test/cryptotest.mak b/src/cpcrypto/Test/cryptotest.mak
deleted file mode 100755
index 8cd82f2..0000000
diff --git a/src/cpcrypto/Test/makealltests.sh b/src/cpcrypto/Test/makealltests.sh
deleted file mode 100755
index be5594c..0000000
diff --git a/src/cpcrypto/Test/mpTest.cc b/src/cpcrypto/Test/mpTest.cc
deleted file mode 100644
index 621eeec..0000000
diff --git a/src/cpcrypto/Test/mpTest.mak b/src/cpcrypto/Test/mpTest.mak
deleted file mode 100644
index f71e82c..0000000
diff --git a/src/cpcrypto/Test/polytest.cc b/src/cpcrypto/Test/polytest.cc
deleted file mode 100755
index bdf4f46..0000000
diff --git a/src/cpcrypto/Test/rsaspeedtest.cc b/src/cpcrypto/Test/rsaspeedtest.cc
deleted file mode 100644
index 35c613d..0000000
diff --git a/src/cpcrypto/Test/rsaspeedtest.mak b/src/cpcrypto/Test/rsaspeedtest.mak
deleted file mode 100644
index 618e164..0000000
diff --git a/src/cpcrypto/Test/sha256speedtest.cc b/src/cpcrypto/Test/sha256speedtest.cc
deleted file mode 100755
index 214253b..0000000
diff --git a/src/cpcrypto/Test/sha256speedtest.mak b/src/cpcrypto/Test/sha256speedtest.mak
deleted file mode 100755
index 8a22fb2..0000000
diff --git a/src/cpcrypto/Test/sha256test.cc b/src/cpcrypto/Test/sha256test.cc
deleted file mode 100755
index caba578..0000000
diff --git a/src/cpcrypto/Test/shatest.cc b/src/cpcrypto/Test/shatest.cc
deleted file mode 100755
index 8991e77..0000000
diff --git a/src/cpcrypto/Test/timing.cc b/src/cpcrypto/Test/timing.cc
deleted file mode 100755
index b934170..0000000
diff --git a/src/cpcrypto/Test/timing.h b/src/cpcrypto/Test/timing.h
deleted file mode 100755
index 79c53fe..0000000
diff --git a/src/cpcrypto/algs.h b/src/cpcrypto/algs.h
deleted file mode 100644
index 38398c7..0000000
diff --git a/src/cpcrypto/bignum/DIRECTORYNOTICE.txt b/src/cpcrypto/bignum/DIRECTORYNOTICE.txt
deleted file mode 100644
index d50133d..0000000
diff --git a/src/cpcrypto/bignum/bignum.h b/src/cpcrypto/bignum/bignum.h
deleted file mode 100755
index 03a75a0..0000000
diff --git a/src/cpcrypto/bignum/fastArith.cc b/src/cpcrypto/bignum/fastArith.cc
deleted file mode 100644
index f8bb194..0000000
diff --git a/src/cpcrypto/bignum/fastArith.h b/src/cpcrypto/bignum/fastArith.h
deleted file mode 100644
index 6dc74ec..0000000
diff --git a/src/cpcrypto/bignum/mpBasicArith.cc b/src/cpcrypto/bignum/mpBasicArith.cc
deleted file mode 100644
index e7714d9..0000000
diff --git a/src/cpcrypto/bignum/mpFunctions.h b/src/cpcrypto/bignum/mpFunctions.h
deleted file mode 100644
index 92a6a8e..0000000
diff --git a/src/cpcrypto/bignum/mpModArith.cc b/src/cpcrypto/bignum/mpModArith.cc
deleted file mode 100644
index 71fe262..0000000
diff --git a/src/cpcrypto/bignum/mpNumTheory.cc b/src/cpcrypto/bignum/mpNumTheory.cc
deleted file mode 100644
index 712f6ea..0000000
diff --git a/src/cpcrypto/bignum/mpRand.cc b/src/cpcrypto/bignum/mpRand.cc
deleted file mode 100644
index 65156e8..0000000
diff --git a/src/cpcrypto/common.h b/src/cpcrypto/common.h
deleted file mode 100644
index debd69e..0000000
diff --git a/src/cpcrypto/cpcryptolib.mak b/src/cpcrypto/cpcryptolib.mak
deleted file mode 100644
index 8376299..0000000
diff --git a/src/cpcrypto/ecc/ecc.h b/src/cpcrypto/ecc/ecc.h
deleted file mode 100755
index 196fc6d..0000000
diff --git a/src/cpcrypto/ecc/ecccrypt.cc b/src/cpcrypto/ecc/ecccrypt.cc
deleted file mode 100644
index ace28d1..0000000
diff --git a/src/cpcrypto/ecc/eccops.cc b/src/cpcrypto/ecc/eccops.cc
deleted file mode 100644
index 5c8e28e..0000000
diff --git a/src/cpcrypto/ecc/nist.cc b/src/cpcrypto/ecc/nist.cc
deleted file mode 100755
index 531b813..0000000
diff --git a/src/cpcrypto/ecc/nist.h b/src/cpcrypto/ecc/nist.h
deleted file mode 100755
index 905e234..0000000
diff --git a/src/cpcrypto/ecc/pointcount/bsgs.cc b/src/cpcrypto/ecc/pointcount/bsgs.cc
deleted file mode 100644
index 918f78a..0000000
diff --git a/src/cpcrypto/ecc/pointcount/divpolys.cc b/src/cpcrypto/ecc/pointcount/divpolys.cc
deleted file mode 100644
index 9ebdc9b..0000000
diff --git a/src/cpcrypto/ecc/pointcount/pointcount.h b/src/cpcrypto/ecc/pointcount/pointcount.h
deleted file mode 100755
index a8edf7f..0000000
diff --git a/src/cpcrypto/ecc/pointcount/polyarith.cc b/src/cpcrypto/ecc/pointcount/polyarith.cc
deleted file mode 100644
index 197c95f..0000000
diff --git a/src/cpcrypto/ecc/pointcount/polyarith.h b/src/cpcrypto/ecc/pointcount/polyarith.h
deleted file mode 100644
index 31f4dfa..0000000
diff --git a/src/cpcrypto/ecc/pointcount/schoof.cc b/src/cpcrypto/ecc/pointcount/schoof.cc
deleted file mode 100644
index 7cbd556..0000000
diff --git a/src/cpcrypto/ecc/pointcount/schoof.cc.old b/src/cpcrypto/ecc/pointcount/schoof.cc.old
deleted file mode 100644
index 8eb7b80..0000000
diff --git a/src/cpcrypto/support/DIRECTORYNOTICE.txt b/src/cpcrypto/support/DIRECTORYNOTICE.txt
deleted file mode 100644
index d50133d..0000000
diff --git a/src/cpcrypto/support/cryptoHelper.cc b/src/cpcrypto/support/cryptoHelper.cc
deleted file mode 100644
index a59914e..0000000
diff --git a/src/cpcrypto/support/cryptoHelper.h b/src/cpcrypto/support/cryptoHelper.h
deleted file mode 100644
index beea3a7..0000000
diff --git a/src/cpcrypto/support/jlmUtility.cc b/src/cpcrypto/support/jlmUtility.cc
deleted file mode 100644
index f9d678b..0000000
diff --git a/src/cpcrypto/support/jlmUtility.h b/src/cpcrypto/support/jlmUtility.h
deleted file mode 100644
index 5cfc07d..0000000
diff --git a/src/cpcrypto/support/jlmcrypto.cc b/src/cpcrypto/support/jlmcrypto.cc
deleted file mode 100644
index d9d762d..0000000
diff --git a/src/cpcrypto/support/jlmcrypto.h b/src/cpcrypto/support/jlmcrypto.h
deleted file mode 100644
index c4e7786..0000000
diff --git a/src/cpcrypto/support/keys.cc b/src/cpcrypto/support/keys.cc
deleted file mode 100644
index 670abc2..0000000
diff --git a/src/cpcrypto/support/keys.h b/src/cpcrypto/support/keys.h
deleted file mode 100644
index 726a29c..0000000
diff --git a/src/cpcrypto/support/logging.cc b/src/cpcrypto/support/logging.cc
deleted file mode 100644
index b804318..0000000
diff --git a/src/cpcrypto/support/logging.h b/src/cpcrypto/support/logging.h
deleted file mode 100644
index d468525..0000000
diff --git a/src/cpcrypto/support/timer.h b/src/cpcrypto/support/timer.h
deleted file mode 100644
index 986529f..0000000
diff --git a/src/cpcrypto/support/tinystr.cc b/src/cpcrypto/support/tinystr.cc
deleted file mode 100755
index 0665768..0000000
diff --git a/src/cpcrypto/support/tinystr.h b/src/cpcrypto/support/tinystr.h
deleted file mode 100755
index 89cca33..0000000
diff --git a/src/cpcrypto/support/tinyxml.cc b/src/cpcrypto/support/tinyxml.cc
deleted file mode 100755
index 1a4c9f3..0000000
diff --git a/src/cpcrypto/support/tinyxml.h b/src/cpcrypto/support/tinyxml.h
deleted file mode 100755
index a3589e5..0000000
diff --git a/src/cpcrypto/support/tinyxmlerror.cc b/src/cpcrypto/support/tinyxmlerror.cc
deleted file mode 100755
index 538c21d..0000000
diff --git a/src/cpcrypto/support/tinyxmlparser.cc b/src/cpcrypto/support/tinyxmlparser.cc
deleted file mode 100755
index 81b7eae..0000000
diff --git a/src/cpcrypto/symmetric/DIRECTORYNOTICE.txt b/src/cpcrypto/symmetric/DIRECTORYNOTICE.txt
deleted file mode 100644
index d50133d..0000000
diff --git a/src/cpcrypto/symmetric/aes.cc b/src/cpcrypto/symmetric/aes.cc
deleted file mode 100755
index 655c9b3..0000000
diff --git a/src/cpcrypto/symmetric/aes.h b/src/cpcrypto/symmetric/aes.h
deleted file mode 100755
index ff5952f..0000000
diff --git a/src/cpcrypto/symmetric/aesni.cc b/src/cpcrypto/symmetric/aesni.cc
deleted file mode 100755
index 9dc3226..0000000
diff --git a/src/cpcrypto/symmetric/aesni.h b/src/cpcrypto/symmetric/aesni.h
deleted file mode 100755
index 1aa4fa7..0000000
diff --git a/src/cpcrypto/symmetric/hmacsha256.cc b/src/cpcrypto/symmetric/hmacsha256.cc
deleted file mode 100644
index dcae1df..0000000
diff --git a/src/cpcrypto/symmetric/hmacsha256.h b/src/cpcrypto/symmetric/hmacsha256.h
deleted file mode 100644
index a213ee6..0000000
diff --git a/src/cpcrypto/symmetric/modesandpadding.cc b/src/cpcrypto/symmetric/modesandpadding.cc
deleted file mode 100644
index 5ac26d7..0000000
diff --git a/src/cpcrypto/symmetric/modesandpadding.h b/src/cpcrypto/symmetric/modesandpadding.h
deleted file mode 100644
index bc45441..0000000
diff --git a/src/cpcrypto/symmetric/sha1.cc b/src/cpcrypto/symmetric/sha1.cc
deleted file mode 100644
index fe9e7b8..0000000
diff --git a/src/cpcrypto/symmetric/sha1.h b/src/cpcrypto/symmetric/sha1.h
deleted file mode 100755
index 6b716cc..0000000
diff --git a/src/cpcrypto/symmetric/sha256.cc b/src/cpcrypto/symmetric/sha256.cc
deleted file mode 100644
index 7d1a695..0000000
diff --git a/src/cpcrypto/symmetric/sha256.h b/src/cpcrypto/symmetric/sha256.h
deleted file mode 100755
index 0328ac1..0000000
diff --git a/src/cpcrypto/symmetric/testBase64.cc b/src/cpcrypto/symmetric/testBase64.cc
deleted file mode 100644
index eb8ae29..0000000
diff --git a/src/tao/CMakeLists.txt b/src/tao/CMakeLists.txt
index 8a46ddd..2eba5db 100644
--- a/src/tao/CMakeLists.txt
+++ b/src/tao/CMakeLists.txt
@@ -13,80 +13,31 @@
 # limitations under the License.
 
 set(TAO_PROTO
-    acl_guard.proto
-    attestation.proto
-    datalog_guard.proto
-    keys.proto
-    linux_admin_rpc.proto
-    linux_host.proto
-    log_net.proto
-    tao_domain.proto
     tao_rpc.proto
    )
 
 PROTOBUF_GENERATE_CPP(PROTO_SRCS PROTO_HDRS ${TAO_PROTO})
 
 set(TAO_SOURCES
-    acl_guard.cc
-    attestation.cc
-    auth.cc
-    datalog_guard.cc
     fd_message_channel.cc
-    keys.cc
-    linux_admin_rpc.cc
-    linux_host.cc
-    linux_process_factory.cc
-    log_net.cc
     message_channel.cc
-    pipe_factory.cc
-    soft_tao.cc
-    tao.cc
-    tao_domain.cc
-    tao_guard.cc
-    tao_root_host.cc
     tao_rpc.cc
-    tao_stacked_host.cc
-    tpm_tao.cc
-    unix_socket_factory.cc
     util.cc
    )
 
 set(TAO_HEADERS
-    acl_guard.h
-    attestation.h
-    auth.h
-    datalog_guard.h
     fd_message_channel.h
-    keys.h
-    linux_admin_rpc.h
-    linux_host.h
-    linux_process_factory.h
-    log_net.h
     message_channel.h
-    pipe_factory.h
-    soft_tao.h
     tao.h
-    tao_domain.h
-    tao_guard.h
-    tao_host.h
-    tao_root_host.h
     tao_rpc.h
-    tao_stacked_host.h
-    tpm_tao.h
-    trivial_guard.h
-    unix_socket_factory.h
     util.h
    )
 
-add_lua_header(auth_lua.h auth.lua)
-
 # This is added to get the right path for the protobuf generated header files.
 include_directories(${CMAKE_BINARY_DIR})
 include_directories(${CMAKE_SOURCE_DIR})
 include_directories(${CMAKE_SOURCE_DIR}/third_party/google-glog/src)
 include_directories(${CMAKE_SOURCE_DIR}/third_party/gflags/src)
-include_directories(${CMAKE_SOURCE_DIR}/third_party/datalog)
-include_directories(${CMAKE_SOURCE_DIR}/third_party/datalog/lua)
 
 # for the chromium definitions, we need -DOS_POSIX
 add_definitions(-DOS_POSIX)
@@ -98,13 +49,11 @@ add_library(tao STATIC
     ${PROTO_HDRS}
     ${TAO_SOURCES}
     ${TAO_HEADERS}
-    ${CMAKE_CURRENT_BINARY_DIR}/auth_lua.h
    )
 
 # TODO(tmroeder): add checks for OpenSSL, libvirt, and libtspi.
 target_link_libraries(tao
     chromium
-    datalog
     gflags
     glog
     modp
@@ -114,29 +63,3 @@ target_link_libraries(tao
     virt
     tspi
    )
-
-# Test code.
-set(TEST_SOURCES
-    attestation_unittests.cc
-    auth_unittests.cc
-    datalog_guard_unittests.cc
-    keys_unittests.cc
-    linux_host_unittests.cc
-    linux_process_factory_unittests.cc
-    pipe_factory_unittests.cc
-    tao_domain_unittests.cc
-    tao_test.cc
-    tao_unittests.cc
-    unix_socket_factory_unittests.cc
-    util_unittests.cc
-   )
-
-set(TEST_HEADERS
-    tao_test.h
-   )
-
-include_directories(../third_party/googlemock/gtest/include)
-add_executable(tao_test ${TEST_SOURCES} ${TEST_HEADERS})
-target_link_libraries(tao_test gtest_main)
-target_link_libraries(tao_test tao)
-add_test(tao_test tao)
diff --git a/src/tao/acl_guard.cc b/src/tao/acl_guard.cc
deleted file mode 100644
index ef39a46..0000000
diff --git a/src/tao/acl_guard.h b/src/tao/acl_guard.h
deleted file mode 100644
index d89ea42..0000000
diff --git a/src/tao/acl_guard.proto b/src/tao/acl_guard.proto
deleted file mode 100644
index a0dab27..0000000
diff --git a/src/tao/attestation.cc b/src/tao/attestation.cc
deleted file mode 100644
index 0d33ad9..0000000
diff --git a/src/tao/attestation.h b/src/tao/attestation.h
deleted file mode 100644
index 50a4c1a..0000000
diff --git a/src/tao/attestation.proto b/src/tao/attestation.proto
deleted file mode 100644
index 0b4a3bb..0000000
diff --git a/src/tao/attestation_unittests.cc b/src/tao/attestation_unittests.cc
deleted file mode 100644
index 45842fb..0000000
diff --git a/src/tao/auth.cc b/src/tao/auth.cc
deleted file mode 100644
index e740a67..0000000
diff --git a/src/tao/auth.h b/src/tao/auth.h
deleted file mode 100644
index 52b8c88..0000000
diff --git a/src/tao/auth.lua b/src/tao/auth.lua
deleted file mode 100644
index f343811..0000000
diff --git a/src/tao/auth_unittests.cc b/src/tao/auth_unittests.cc
deleted file mode 100644
index b90e7dc..0000000
diff --git a/src/tao/datalog_guard.cc b/src/tao/datalog_guard.cc
deleted file mode 100644
index 8db5a9e..0000000
diff --git a/src/tao/datalog_guard.h b/src/tao/datalog_guard.h
deleted file mode 100644
index 0dc2a76..0000000
diff --git a/src/tao/datalog_guard.proto b/src/tao/datalog_guard.proto
deleted file mode 100644
index bd6baf8..0000000
diff --git a/src/tao/datalog_guard_unittests.cc b/src/tao/datalog_guard_unittests.cc
deleted file mode 100644
index 2540671..0000000
diff --git a/src/tao/keys.cc b/src/tao/keys.cc
deleted file mode 100644
index 0d2b9f1..0000000
diff --git a/src/tao/keys.h b/src/tao/keys.h
deleted file mode 100644
index dafeca0..0000000
diff --git a/src/tao/keys.proto b/src/tao/keys.proto
deleted file mode 100644
index 8d302e3..0000000
diff --git a/src/tao/keys_unittests.cc b/src/tao/keys_unittests.cc
deleted file mode 100644
index 5678963..0000000
diff --git a/src/tao/linux_admin_rpc.cc b/src/tao/linux_admin_rpc.cc
deleted file mode 100644
index ce9a7cd..0000000
diff --git a/src/tao/linux_admin_rpc.h b/src/tao/linux_admin_rpc.h
deleted file mode 100644
index b1dcc9c..0000000
diff --git a/src/tao/linux_admin_rpc.proto b/src/tao/linux_admin_rpc.proto
deleted file mode 100644
index d6119b7..0000000
diff --git a/src/tao/linux_host.cc b/src/tao/linux_host.cc
deleted file mode 100644
index 254f8d0..0000000
diff --git a/src/tao/linux_host.h b/src/tao/linux_host.h
deleted file mode 100644
index 7348f58..0000000
diff --git a/src/tao/linux_host.proto b/src/tao/linux_host.proto
deleted file mode 100644
index a8505f3..0000000
diff --git a/src/tao/linux_host_unittests.cc b/src/tao/linux_host_unittests.cc
deleted file mode 100644
index ea69200..0000000
diff --git a/src/tao/linux_process_factory.cc b/src/tao/linux_process_factory.cc
deleted file mode 100644
index 3f274ca..0000000
diff --git a/src/tao/linux_process_factory.h b/src/tao/linux_process_factory.h
deleted file mode 100644
index c667393..0000000
diff --git a/src/tao/linux_process_factory_unittests.cc b/src/tao/linux_process_factory_unittests.cc
deleted file mode 100644
index e999a7c..0000000
diff --git a/src/tao/log_net.cc b/src/tao/log_net.cc
deleted file mode 100644
index b0ca99a..0000000
diff --git a/src/tao/log_net.h b/src/tao/log_net.h
deleted file mode 100644
index 3c49c32..0000000
diff --git a/src/tao/log_net.proto b/src/tao/log_net.proto
deleted file mode 100644
index 8e86a4c..0000000
diff --git a/src/tao/message_channel.cc b/src/tao/message_channel.cc
index c9b7da6..a16df85 100644
--- a/src/tao/message_channel.cc
+++ b/src/tao/message_channel.cc
@@ -25,6 +25,7 @@
 #include <string>
 
 #include <glog/logging.h>
+#include <google/protobuf/message.h>
 
 #include "tao/util.h"
 
diff --git a/src/tao/pipe_factory.cc b/src/tao/pipe_factory.cc
deleted file mode 100644
index c1fc2f8..0000000
diff --git a/src/tao/pipe_factory.h b/src/tao/pipe_factory.h
deleted file mode 100644
index 6250d8b..0000000
diff --git a/src/tao/pipe_factory_unittests.cc b/src/tao/pipe_factory_unittests.cc
deleted file mode 100644
index 5524f75..0000000
diff --git a/src/tao/soft_tao.cc b/src/tao/soft_tao.cc
deleted file mode 100644
index 75411e6..0000000
diff --git a/src/tao/soft_tao.h b/src/tao/soft_tao.h
deleted file mode 100644
index ba3affc..0000000
diff --git a/src/tao/tao.cc b/src/tao/tao.cc
deleted file mode 100644
index ede9491..0000000
diff --git a/src/tao/tao.h b/src/tao/tao.h
index 30a8c22..ae89e31 100644
--- a/src/tao/tao.h
+++ b/src/tao/tao.h
@@ -21,8 +21,6 @@
 
 #include <string>
 
-#include "tao/attestation.pb.h"
-
 namespace tao {
 using std::string;
 
@@ -38,9 +36,7 @@ using std::string;
 class Tao {
  public:
   Tao() {}
-  virtual ~Tao() {
-    if (host_tao_ == this) host_tao_ = nullptr;
-  }
+  virtual ~Tao() {}
 
   /// Serialize Tao parameters for passing across fork/exec or between
   /// processes, if possible. Not all Tao implementations must necessarily be
@@ -48,12 +44,6 @@ class Tao {
   /// @param params[out] The serialized parameters.
   virtual bool SerializeToString(string *params) const { return false; }
 
-  /// Get the interface to the host Tao for this hosted program.
-  /// Every hosted program is provided with one such interface.
-  /// Ownership is of the returned pointer retained by this (static) class,
-  /// though the caller can do a Close() if desired.
-  static Tao *GetHostTao();
-
   /// Get the Tao principal name assigned to this hosted program. The name
   /// encodes the full path from the root Tao, through all intermediary Tao
   /// hosts, to this hosted program. The name will be globally unique: different
@@ -108,10 +98,11 @@ class Tao {
   /// @}
 
   /// Request the Tao host sign a Statement on behalf of this hosted program.
-  /// @param stmt A Statement to be signed. The issuer, time, and expiration
-  /// fields will be filled in with appropriate defaults if they are left empty.
+  /// @param message A delegation statement to be signed. This statement is a
+  /// serialized binary auth statement and can be created using MarshalSpeaksfor
+  /// in util.h.
   /// @param[out] attestation The resulting signed attestation.
-  virtual bool Attest(const Statement &stmt, string *attestation) = 0;
+  virtual bool Attest(const string &message, string *attestation) = 0;
 
   /// Encrypt data so only certain hosted programs can unseal it.
   /// @param data The data to seal.
@@ -171,13 +162,6 @@ class Tao {
 
   /// Default timeout for Attestation (= 1 year in seconds).
   static const int DefaultAttestationTimeout = 31556926;
-
-  /// Environment variable for passing parent/child channel parameters from a
-  /// parent to a hosted process.
-  constexpr static auto HostTaoEnvVar = "GOOGLE_HOST_TAO";
-
- private:
-  static Tao *host_tao_;
 };
 }  // namespace tao
 
diff --git a/src/tao/tao_domain.cc b/src/tao/tao_domain.cc
deleted file mode 100644
index d943860..0000000
diff --git a/src/tao/tao_domain.h b/src/tao/tao_domain.h
deleted file mode 100644
index 3032090..0000000
diff --git a/src/tao/tao_domain.proto b/src/tao/tao_domain.proto
deleted file mode 100644
index 5122beb..0000000
diff --git a/src/tao/tao_domain_unittests.cc b/src/tao/tao_domain_unittests.cc
deleted file mode 100644
index 42c53c1..0000000
diff --git a/src/tao/tao_guard.cc b/src/tao/tao_guard.cc
deleted file mode 100644
index ef18155..0000000
diff --git a/src/tao/tao_guard.h b/src/tao/tao_guard.h
deleted file mode 100644
index 4258368..0000000
diff --git a/src/tao/tao_host.h b/src/tao/tao_host.h
deleted file mode 100644
index 9168ba3..0000000
diff --git a/src/tao/tao_root_host.cc b/src/tao/tao_root_host.cc
deleted file mode 100644
index 41fe8a6..0000000
diff --git a/src/tao/tao_root_host.h b/src/tao/tao_root_host.h
deleted file mode 100644
index 5126f4e..0000000
diff --git a/src/tao/tao_rpc.cc b/src/tao/tao_rpc.cc
index 0c7a678..ba34d79 100644
--- a/src/tao/tao_rpc.cc
+++ b/src/tao/tao_rpc.cc
@@ -49,14 +49,9 @@ bool TaoRPC::GetSharedSecret(size_t size, const string &policy, string *bytes) {
   return Request("Tao.GetSharedSecret", rpc, bytes, nullptr /* policy */);
 }
 
-bool TaoRPC::Attest(const Statement &stmt, string *attestation) {
-  string serialized_stmt;
-  if (!stmt.SerializePartialToString(&serialized_stmt)) {
-    LOG(ERROR) << "Could not serialize partial statement";
-    return false;
-  }
+bool TaoRPC::Attest(const string &message, string *attestation) {
   TaoRPCRequest rpc;
-  rpc.set_data(serialized_stmt);
+  rpc.set_data(message);
   return Request("Tao.Attest", rpc, attestation, nullptr /* policy */);
 }
 
diff --git a/src/tao/tao_rpc.h b/src/tao/tao_rpc.h
index a81063b..772263c 100644
--- a/src/tao/tao_rpc.h
+++ b/src/tao/tao_rpc.h
@@ -50,7 +50,7 @@ class TaoRPC : public Tao {
   virtual bool GetRandomBytes(size_t size, string *bytes);
   virtual bool GetSharedSecret(size_t size, const string &policy,
                                string *bytes);
-  virtual bool Attest(const Statement &stmt, string *attestation);
+  virtual bool Attest(const string &message, string *attestation);
   virtual bool Seal(const string &data, const string &policy, string *sealed);
   virtual bool Unseal(const string &sealed, string *data, string *policy);
   virtual string GetRecentErrorMessage() const { return failure_msg_; }
diff --git a/src/tao/tao_rpc.proto b/src/tao/tao_rpc.proto
index a58e087..16aef20 100644
--- a/src/tao/tao_rpc.proto
+++ b/src/tao/tao_rpc.proto
@@ -19,7 +19,7 @@
 
 package tao;
 
-// Copied from: go/src/cloudproxy/util/protorpc/protorpc.proto
+// Copied from: util/protorpc/protorpc.proto
 
 // Protobuf RPC request header
 message ProtoRPCRequestHeader {
@@ -42,12 +42,15 @@ message ProtoRPCResponseHeader {
   optional string error = 3;
 }
 
-// Copied from: go/src/cloudproxy/tao/tao_rpc.proto
+// Copied from: tao/tao_rpc.proto
 
 message TaoRPCRequest {
   optional bytes data = 1;
   optional int32 size = 2;
   optional string policy = 3;
+  optional int64 time = 4;
+  optional int64 expiration = 5;
+  optional bytes issuer = 6;
 }
 
 message TaoRPCResponse {
diff --git a/src/tao/tao_stacked_host.cc b/src/tao/tao_stacked_host.cc
deleted file mode 100644
index e32ea37..0000000
diff --git a/src/tao/tao_stacked_host.h b/src/tao/tao_stacked_host.h
deleted file mode 100644
index 2a499b8..0000000
diff --git a/src/tao/tao_test.cc b/src/tao/tao_test.cc
deleted file mode 100644
index 6099b2f..0000000
diff --git a/src/tao/tao_test.h b/src/tao/tao_test.h
deleted file mode 100644
index d9a7ed0..0000000
diff --git a/src/tao/tao_unittests.cc b/src/tao/tao_unittests.cc
deleted file mode 100644
index b6e1b68..0000000
diff --git a/src/tao/tpm_tao.cc b/src/tao/tpm_tao.cc
deleted file mode 100644
index a06c180..0000000
diff --git a/src/tao/tpm_tao.h b/src/tao/tpm_tao.h
deleted file mode 100644
index e23635b..0000000
diff --git a/src/tao/trivial_guard.h b/src/tao/trivial_guard.h
deleted file mode 100644
index 87aedf7..0000000
diff --git a/src/tao/unix_socket_factory.cc b/src/tao/unix_socket_factory.cc
deleted file mode 100644
index 6d8418f..0000000
diff --git a/src/tao/unix_socket_factory.h b/src/tao/unix_socket_factory.h
deleted file mode 100644
index 4fe6859..0000000
diff --git a/src/tao/unix_socket_factory_unittests.cc b/src/tao/unix_socket_factory_unittests.cc
deleted file mode 100644
index 09705e0..0000000
diff --git a/src/tao/util.cc b/src/tao/util.cc
index bcb51ba..3e8cb37 100644
--- a/src/tao/util.cc
+++ b/src/tao/util.cc
@@ -38,6 +38,8 @@
 
 #include <gflags/gflags.h>
 #include <glog/logging.h>
+#include <google/protobuf/io/coded_stream.h>
+#include <google/protobuf/io/zero_copy_stream_impl_lite.h>
 #include <modp/modp_b64w.h>
 #include <openssl/bio.h>
 #include <openssl/crypto.h>
@@ -45,11 +47,7 @@
 #include <openssl/rand.h>
 #include <openssl/ssl.h>
 
-#include "tao/attestation.pb.h"
-#include "tao/keys.h"
-#include "tao/log_net.h"
 #include "tao/tao.h"
-#include "tao/tao_domain.h"
 
 using std::lock_guard;
 using std::mutex;
@@ -57,6 +55,9 @@ using std::shared_ptr;
 using std::stringstream;
 using std::vector;
 
+using google::protobuf::io::StringOutputStream;
+using google::protobuf::io::CodedOutputStream;
+
 namespace tao {
 
 void SecureStringErase(string *s) {
@@ -103,43 +104,6 @@ static void locking_function(int mode, int n, const char *file, int line) {
   }
 }
 
-bool Sha256FileHash(const string &path, string *hash) {
-  string contents;
-  if (!ReadFileToString(path, &contents)) {
-    LOG(ERROR) << "Can't read " << path;
-    return false;
-  }
-  return Sha256(contents, hash);
-}
-
-bool Sha1FileHash(const string &path, string *hash) {
-  string contents;
-  if (!ReadFileToString(path, &contents)) {
-    LOG(ERROR) << "Can't read " << path;
-    return false;
-  }
-  return Sha1(contents, hash);
-}
-
-/*
-bool RegisterKnownChannels(TaoChildChannelRegistry *registry) {
-  if (registry == nullptr) {
-    LOG(ERROR) << "Could not register channels with a null registry";
-    return false;
-  }
-
-  registry->Register(
-      KvmUnixTaoChildChannel::ChannelType(),
-      TaoChildChannelRegistry::CallConstructor<KvmUnixTaoChildChannel>);
-
-  registry->Register(
-      PipeTaoChildChannel::ChannelType(),
-      TaoChildChannelRegistry::CallConstructor<PipeTaoChildChannel>);
-
-  return true;
-}
-*/
-
 bool OpenSSLSuccess() {
   uint32_t last_error = ERR_get_error();
   if (last_error) {
@@ -182,179 +146,10 @@ bool InitializeApp(int *argc, char ***argv, bool remove_args) {
   google::ParseCommandLineFlags(argc, argv, remove_args);
   google::InitGoogleLogging((*argv)[0]);
   google::InstallFailureSignalHandler();
-  LogNet::Init("localhost", "5514", (*argv)[0]);
   signal(SIGPIPE, SIG_IGN);
   return InitializeOpenSSL();
 }
 
-constexpr int MaxSelfPipeSignum = NSIG;
-struct SelfPipe {
-  bool open;
-  int fd[2];
-  struct sigaction sa;
-};
-SelfPipe selfPipe[MaxSelfPipeSignum] = {};
-static mutex selfPipeMutex;
-
-static void SelfPipeHandler(int signum) {
-  if (signum <= 0 || signum > MaxSelfPipeSignum) return;
-  if (!selfPipe[signum - 1].open) return;
-  int savedErrno = errno;
-  char b = static_cast<char>(signum);
-  ssize_t v = write(selfPipe[signum - 1].fd[1], &b, 1);
-  (void) v;  // Ignore the result
-  errno = savedErrno;
-}
-
-static bool SetFdFlags(int fd, int flags) {
-  int f = fcntl(fd, F_GETFL);
-  if (f == -1) {
-    LOG(ERROR) << "Could not get flags for fd " << fd;
-    return false;
-  }
-  if (fcntl(fd, F_SETFL, f | flags) == -1) {
-    LOG(ERROR) << "Could not set flags for fd " << fd;
-    return false;
-  }
-  return true;
-}
-
-int GetSelfPipeSignalFd(int signum, int sa_flags) {
-  if (signum <= 0 || signum > MaxSelfPipeSignum) {
-    LOG(ERROR) << "Invalid self-pipe signal number " << signum;
-    return -1;
-  }
-  lock_guard<mutex> l(selfPipeMutex);
-  if (selfPipe[signum - 1].open) {
-    LOG(ERROR) << "Self-pipe already opened";
-    // We could instead return the existing fd here if callers can share it.
-    return -1;
-  }
-  if (pipe(selfPipe[signum - 1].fd) == -1) {
-    LOG(ERROR) << "Could not create self-pipe";
-    return -1;
-  }
-  if (!SetFdFlags(selfPipe[signum - 1].fd[0], O_NONBLOCK) ||
-      !SetFdFlags(selfPipe[signum - 1].fd[1], O_NONBLOCK)) {
-    PLOG(ERROR) << "Could not set self-pipe disposition";
-    close(selfPipe[signum - 1].fd[0]);
-    close(selfPipe[signum - 1].fd[1]);
-    selfPipe[signum - 1].fd[0] = selfPipe[signum - 1].fd[1] = -1;
-    return -1;
-  }
-  struct sigaction act;
-  memset(&act, 0, sizeof(struct sigaction));
-  act.sa_handler = SelfPipeHandler;
-  act.sa_flags = sa_flags;
-  if (sigaction(signum, &act, &selfPipe[signum - 1].sa) < 0) {
-    PLOG(ERROR) << "Could not set self-pipe handler";
-    close(selfPipe[signum - 1].fd[0]);
-    close(selfPipe[signum - 1].fd[1]);
-    selfPipe[signum - 1].fd[0] = selfPipe[signum - 1].fd[1] = -1;
-    return -1;
-  }
-  selfPipe[signum - 1].open = true;
-  return selfPipe[signum - 1].fd[0];
-}
-
-bool ReleaseSelfPipeSignalFd(int fd) {
-  if (fd < 0) {
-    LOG(ERROR) << "Invalid self-pipe fd " << fd;
-    return false;
-  }
-  lock_guard<mutex> l(selfPipeMutex);
-  for (int signum = 1; signum <= MaxSelfPipeSignum; signum++) {
-    if (!selfPipe[signum - 1].open || selfPipe[signum - 1].fd[0] != fd)
-      continue;
-    selfPipe[signum - 1].open = false;
-    bool success = true;
-    if (sigaction(signum, &selfPipe[signum - 1].sa, nullptr) < 0) {
-      PLOG(ERROR) << "Could not restore old handler for signal " << signum;
-      success = false;
-    }
-    close(selfPipe[signum - 1].fd[0]);
-    close(selfPipe[signum - 1].fd[1]);
-    selfPipe[signum - 1].fd[0] = selfPipe[signum - 1].fd[1] = -1;
-    return success;
-  }
-  LOG(ERROR) << "No such self-pipe fd " << fd;
-  return false;
-}
-
-void selfpipe_release(int *fd) {
-  if (fd && *fd >= 0) {
-    if (!ReleaseSelfPipeSignalFd(*fd)) {
-      PLOG(ERROR) << "Could not close self-pipe fd " << *fd;
-    }
-    delete fd;
-  }
-}
-
-bool OpenTCPSocket(const string &host, const string &port, int *sock) {
-  if (sock == nullptr) {
-    LOG(ERROR) << "null socket parameter";
-    return false;
-  }
-
-  *sock = socket(AF_INET, SOCK_STREAM, 0);
-  if (*sock == -1) {
-    PLOG(ERROR) << "Could not create a socket to listen on";
-    return false;
-  }
-
-  // Don't allow TIME_WAIT sockets from interfering with bind() below. The
-  // socket option SO_REUSEADDR allows this socket to bind even when there is
-  // another bound socket in TIME_WAIT.
-  int val = 1;
-  if (setsockopt(*sock, SOL_SOCKET, SO_REUSEADDR, &val, sizeof(val)) != 0) {
-    PLOG(ERROR) << "Could not set SO_REUSEADDR on the socket for " << host
-                << ":" << port;
-    return false;
-  }
-
-  struct addrinfo hints;
-  memset(&hints, 0, sizeof(struct addrinfo));
-  hints.ai_family = AF_INET;
-  hints.ai_socktype = SOCK_STREAM;
-
-  struct addrinfo *addrs = nullptr;
-  int info_err = getaddrinfo(host.c_str(), port.c_str(), &hints, &addrs);
-  if (info_err != 0) {
-    LOG(ERROR) << "Could not get address information for " << host << ":"
-               << port;
-    return false;
-  }
-
-  int bind_err = bind(*sock, addrs->ai_addr, addrs->ai_addrlen);
-  if (bind_err == -1) {
-    PLOG(ERROR) << "Could not bind the socket";
-    return false;
-  }
-
-  int listen_err = listen(*sock, 128 /* max completed connections */);
-  if (listen_err == -1) {
-    PLOG(ERROR) << "Could not set the socket up for listening";
-    return false;
-  }
-
-  return true;
-}
-
-bool GetTCPSocketInfo(int sock, string *host, string *port) {
-  struct sockaddr_in addr;
-  unsigned int len = sizeof(addr);
-  if (getsockname(sock, (struct sockaddr *)&addr, &len) == -1) {
-    PLOG(ERROR) << "Could not get socket name";
-    return false;
-  }
-  char buf[INET_ADDRSTRLEN];
-  host->assign(inet_ntop(AF_INET, &addr.sin_addr, buf, sizeof(buf)));
-  stringstream out;
-  out << (unsigned)ntohs(addr.sin_port);
-  port->assign(out.str());
-  return true;
-}
-
 bool MakeSealedSecret(Tao *tao, const string &path, const string &policy,
                       int secret_size, string *secret) {
   if (secret == nullptr) {
@@ -672,15 +467,6 @@ bool WeakRandBytes(size_t size, string *s) {
   return (RAND_bytes(str2uchar(s), size) == 1);
 }
 
-bool RandBytes(size_t size, string *s) {
-  Tao *host = Tao::GetHostTao();
-  if (host == nullptr) {
-    return WeakRandBytes(size, s);
-  } else {
-    return host->GetRandomBytes(size, s);
-  }
-}
-
 string Base64WEncode(const string &in) {
   string out;
   Base64WEncode(in, &out);  // does not fail
@@ -713,4 +499,62 @@ bool Base64WDecode(const string &in, string *out) {
   return true;
 }
 
+// These constants give the tags needed to encode a key auth.Prin in binary form
+// in a Speaksfor statement.
+static int tagPrin = 0x1;
+static int tagBytes = 0x4;
+static int tagSpeaksfor = 0xd;
+static int tagSubPrin = 0x11;
+bool MarshalSpeaksfor(const string &key, const string &binaryTaoName,
+                      string *out) {
+
+  StringOutputStream *sos = new StringOutputStream(out);
+  CodedOutputStream *cos = new CodedOutputStream(sos);
+  // Tag this as a Speaksfor object.
+  cos->WriteVarint32(tagSpeaksfor);
+
+  string delegate;
+  if (!MarshalKeyPrin(key, &delegate)) {
+    delete cos;
+    delete sos;
+    return false;
+  }
+
+  cos->WriteRaw(delegate.data(), delegate.size());
+
+  // The delegator is written directly, since it's already a binary-encoded
+  // Term.
+  cos->WriteRaw(binaryTaoName.data(), binaryTaoName.size());
+
+  delete cos;
+  delete sos;
+  return true;
+}
+
+bool MarshalKeyPrin(const string &key, string *out) {
+  StringOutputStream *sos = new StringOutputStream(out);
+  CodedOutputStream *cos = new CodedOutputStream(sos);
+
+  // auth.Prin is tagPrin, then a string type "key", then a Key buffer, and no
+  // extensions (this is marshaled as tagSubPrin, 0).
+  cos->WriteVarint32(tagPrin);
+
+  // Type: "key"
+  string keyType("key");
+  cos->WriteVarint32(keyType.size());
+  cos->WriteRaw(keyType.data(), keyType.size());
+
+  // Key: auth.Bytes[...]
+  cos->WriteVarint32(tagBytes);
+  cos->WriteVarint32(key.size());
+  cos->WriteRaw(key.data(), key.size());
+
+  // Ext: auth.SubPrin of length 0
+  cos->WriteVarint32(tagSubPrin);
+  cos->WriteVarint32(0);
+
+  delete cos;
+  delete sos;
+  return true;
+}
 }  // namespace tao
diff --git a/src/tao/util.h b/src/tao/util.h
index 64da8ca..7126140 100644
--- a/src/tao/util.h
+++ b/src/tao/util.h
@@ -36,8 +36,6 @@
 
 #include "tao/tao.h"
 
-struct sockaddr;
-
 namespace google {
 namespace protobuf {
 class Message;
@@ -82,9 +80,6 @@ std::unique_ptr<T> make_unique(Args &&... args) {
   return std::unique_ptr<T>(new T(std::forward<Args>(args)...));
 }
 
-// class TaoChildChannelRegistry;
-class TaoDomain;
-
 /// Close a file descriptor and ignore the return value. This is used by the
 /// definition of ScopedFd.
 /// @param fd A pointer to the file descriptor to close and free.
@@ -100,11 +95,6 @@ void file_close(FILE *file);
 /// @param dir The path to the directory.
 void temp_file_cleaner(string *dir);
 
-/// Close a self-pipe and ignore the return value. This is used by
-/// the definition of ScopedSelfPipeFd.
-/// @param fd A pointer to the self-pipe file descriptor.
-void selfpipe_release(int *fd);
-
 /// A version of unique_ptr::reset() that returns the new pointer. Useful for
 /// putting reset inside of conditionals.
 /// @param t The unique_ptr to be reset.
@@ -146,9 +136,6 @@ typedef unique_free_ptr<FILE, file_close> ScopedFile;
 /// A smart pointer to a temporary directory to be cleaned upon destruction.
 typedef unique_free_ptr<string, temp_file_cleaner> ScopedTempDir;
 
-/// A smart pointer to a self-pipe.
-typedef unique_free_ptr<int, selfpipe_release> ScopedSelfPipeFd;
-
 /// Extract pointer to string data. These can be used for library functions
 /// that require raw pointers instead of C++ strings. Returned const pointers
 /// should not be written to. Returned non-const pointers can be written. Any
@@ -173,44 +160,6 @@ inline unsigned char *str2uchar(string *s) {
 }
 /// @}
 
-/// Create a self-pipe for a signal. A signal handler is installed that writes
-/// the signal number (cast to a byte) to the pipe. Callers can use the returned
-/// file descriptor as part of a select() call. When a byte is available on the
-/// file descriptor, it means that that signal has been received. An error is
-/// returned if another self-pipe already exists (this limitation stems from the
-/// need for global variables).
-/// @param signum The signal to catch.
-/// @param sa_flags Flags to modify the signal behavior. See sigaction(2).
-/// @return A file descriptor suitable for select() and read(), or -1 on error.
-int GetSelfPipeSignalFd(int signum, int sa_flags);
-
-/// Destroy a self-pipe, restoring any previous signal handler.
-/// @param fd The file descriptor returned from GetSelfPipeSignalFd().
-bool ReleaseSelfPipeSignalFd(int fd);
-
-/// Hash a string using SHA1 or SHA256.
-/// @param s The string to hash.
-/// @param[out] hash The resulting hash.
-/// @{
-bool Sha1(const string &s, string *hash);
-bool Sha256(const string &s, string *hash);
-/// @}
-
-/// Hash a file using SHA1 or SHA256.
-/// @param path The path of the file to hash.
-/// @param[out] hash The resulting hash.
-/// @{
-bool Sha1FileHash(const string &path, string *hash);
-bool Sha256FileHash(const string &path, string *hash);
-/// @}
-
-/// Register some well-known TaoChannels with the registry. The list of
-/// registered TaoChannels is:
-/// - KvmUnixTaoChannel
-/// - PipeTaoChannel
-/// @param registry The registry to fill with the channels
-// bool RegisterKnownChannels(TaoChildChannelRegistry *registry);
-
 /// Call the OpenSSL initialization routines and set up locking for
 /// multi-threaded access.
 bool InitializeOpenSSL();
@@ -245,24 +194,6 @@ bool InitializeApp(int *argc, char ***argv, bool remove_args);
 /// program";
 bool OpenSSLSuccess();
 
-/// Open a listening TCP socket on the given port.
-/// @param host The host to listen on.
-/// @param port The port to listen on.
-/// @param[out] sock The socket opened for this port.
-bool OpenTCPSocket(const string &host, const string &port, int *sock);
-
-/// Get local address information about an open TCP socket.
-/// @param sock The socket.
-/// @param[out] host The local host address.
-/// @param[out] port The local port.
-bool GetTCPSocketInfo(int sock, string *host, string *port);
-
-/// Connect to a remote server.
-/// @param host The name of the remote host.
-/// @param port The port to connect to.
-/// @param[out] sock The connected client socket.
-bool ConnectToTCPServer(const string &host, const string &port, int *sock);
-
 /// Generate and save a random secret, sealed against the host Tao.
 /// @param tao The interface to access the host Tao.
 /// @param path The location to store the sealed secret.
@@ -379,12 +310,6 @@ time_t FileModificationTime(const string &path);
 /// @param[out] s A string in which to place the bytes.
 bool WeakRandBytes(size_t size, string *s);
 
-/// Get random bytes from host Tao, if available, otherwise from OpenSSL.
-/// @param size The number of bytes to get.
-/// @param[out] s A string in which to place the bytes.
-// TODO(kwalsh) Rethink this function.
-bool RandBytes(size_t size, string *s);
-
 /// Encode string using web-safe base64w. No padding or newlines are added. This
 /// function does not fail.
 /// @param in The string to be encoded. May be emptystring.
@@ -403,6 +328,21 @@ bool Base64WEncode(const string &in, string *out);
 /// @param[out] out A string to be overwritten with the result.
 bool Base64WDecode(const string &in, string *out);
 
+/// Encode a key as a binary auth statement that can be parsed by the Go Tao
+/// binary package in github.com/jlmucb/cloudproxy/tao/auth. Note that there is
+/// no corresponding ParseSpeaksfor function, since this code exists only to
+/// pass well-formed messages down to the Tao.
+/// @param key The bytes of the key to authorize.
+/// @param binaryTaoName The binary-marshaled principal name of the delegator.
+/// @param[out] out The resulting binary-serialized auth.Speaksfor statement.
+bool MarshalSpeaksfor(const string &key, const string &binaryTaoName,
+			string *out);
+
+/// Encode a key as an auth.Prin of type "key". This uses the binary encoding
+/// from the Go package github.com/jlmucb/cloudproxy/tao/auth
+/// @param key The bytes of the key to write as an auth.Prin
+/// @param[out] out The resulting binary-encoded auth.Prin.
+bool MarshalKeyPrin(const string &key, string *out);
 }  // namespace tao
 
 #endif  // TAO_UTIL_H_
diff --git a/src/tao/util_unittests.cc b/src/tao/util_unittests.cc
deleted file mode 100644
index 9c78cc0..0000000
diff --git a/src/third_party/CMakeLists.txt b/src/third_party/CMakeLists.txt
index 21721cd..4135a5a 100644
--- a/src/third_party/CMakeLists.txt
+++ b/src/third_party/CMakeLists.txt
@@ -13,7 +13,6 @@
 # limitations under the License.
 
 add_subdirectory(chromium)
-add_subdirectory(datalog)
 add_subdirectory(gflags)
 add_subdirectory(google-glog)
 add_subdirectory(googlemock)
diff --git a/src/third_party/datalog/.deps/bin2c.Po b/src/third_party/datalog/.deps/bin2c.Po
deleted file mode 100644
index 655c32b..0000000
diff --git a/src/third_party/datalog/.deps/datalog.Plo b/src/third_party/datalog/.deps/datalog.Plo
deleted file mode 100644
index 5466a7b..0000000
diff --git a/src/third_party/datalog/.deps/interp.Po b/src/third_party/datalog/.deps/interp.Po
deleted file mode 100644
index 2c17364..0000000
diff --git a/src/third_party/datalog/.deps/loader.Plo b/src/third_party/datalog/.deps/loader.Plo
deleted file mode 100644
index 4e7bef2..0000000
diff --git a/src/third_party/datalog/AUTHORS b/src/third_party/datalog/AUTHORS
deleted file mode 100644
index 3970652..0000000
diff --git a/src/third_party/datalog/CMakeLists.txt b/src/third_party/datalog/CMakeLists.txt
deleted file mode 100644
index 6b15da8..0000000
diff --git a/src/third_party/datalog/COPYING.LIB b/src/third_party/datalog/COPYING.LIB
deleted file mode 100644
index 47044a8..0000000
diff --git a/src/third_party/datalog/ChangeLog b/src/third_party/datalog/ChangeLog
deleted file mode 100644
index 118baba..0000000
diff --git a/src/third_party/datalog/INSTALL b/src/third_party/datalog/INSTALL
deleted file mode 100644
index 6e90e07..0000000
diff --git a/src/third_party/datalog/Makefile.am b/src/third_party/datalog/Makefile.am
deleted file mode 100644
index 7534c18..0000000
diff --git a/src/third_party/datalog/Makefile.in b/src/third_party/datalog/Makefile.in
deleted file mode 100644
index a9f94eb..0000000
diff --git a/src/third_party/datalog/NEWS b/src/third_party/datalog/NEWS
deleted file mode 100644
index 1c5f45d..0000000
diff --git a/src/third_party/datalog/README b/src/third_party/datalog/README
deleted file mode 100644
index 05b28f4..0000000
diff --git a/src/third_party/datalog/acinclude.m4 b/src/third_party/datalog/acinclude.m4
deleted file mode 100644
index f65b35e..0000000
diff --git a/src/third_party/datalog/aclocal.m4 b/src/third_party/datalog/aclocal.m4
deleted file mode 100644
index 66c8f56..0000000
diff --git a/src/third_party/datalog/add.lua b/src/third_party/datalog/add.lua
deleted file mode 100644
index 2d68642..0000000
diff --git a/src/third_party/datalog/ancestor.dl b/src/third_party/datalog/ancestor.dl
deleted file mode 100644
index bf43f97..0000000
diff --git a/src/third_party/datalog/ancestor.txt b/src/third_party/datalog/ancestor.txt
deleted file mode 100644
index 62b3f09..0000000
diff --git a/src/third_party/datalog/bidipath.dl b/src/third_party/datalog/bidipath.dl
deleted file mode 100644
index 3ca9c9a..0000000
diff --git a/src/third_party/datalog/bidipath.txt b/src/third_party/datalog/bidipath.txt
deleted file mode 100644
index 6230a5a..0000000
diff --git a/src/third_party/datalog/bin2c.c b/src/third_party/datalog/bin2c.c
deleted file mode 100644
index c72fbc5..0000000
diff --git a/src/third_party/datalog/config.guess b/src/third_party/datalog/config.guess
deleted file mode 100755
index c0adba9..0000000
diff --git a/src/third_party/datalog/config.log b/src/third_party/datalog/config.log
deleted file mode 100644
index ea789ec..0000000
diff --git a/src/third_party/datalog/config.status b/src/third_party/datalog/config.status
deleted file mode 100755
index 3eaa3eb..0000000
diff --git a/src/third_party/datalog/config.sub b/src/third_party/datalog/config.sub
deleted file mode 100755
index 6205f84..0000000
diff --git a/src/third_party/datalog/configure b/src/third_party/datalog/configure
deleted file mode 100755
index c1e48a6..0000000
diff --git a/src/third_party/datalog/configure.ac b/src/third_party/datalog/configure.ac
deleted file mode 100644
index 4140657..0000000
diff --git a/src/third_party/datalog/datalog.1 b/src/third_party/datalog/datalog.1
deleted file mode 100644
index 97917cf..0000000
diff --git a/src/third_party/datalog/datalog.c b/src/third_party/datalog/datalog.c
deleted file mode 100644
index fcbd1b8..0000000
diff --git a/src/third_party/datalog/datalog.h b/src/third_party/datalog/datalog.h
deleted file mode 100644
index 98940b4..0000000
diff --git a/src/third_party/datalog/datalog.info b/src/third_party/datalog/datalog.info
deleted file mode 100644
index 64cac63..0000000
diff --git a/src/third_party/datalog/datalog.lua b/src/third_party/datalog/datalog.lua
deleted file mode 100644
index 4667ed0..0000000
diff --git a/src/third_party/datalog/datalog.lua.in b/src/third_party/datalog/datalog.lua.in
deleted file mode 100644
index 0d2b22a..0000000
diff --git a/src/third_party/datalog/datalog.spec b/src/third_party/datalog/datalog.spec
deleted file mode 100644
index 7a1e191..0000000
diff --git a/src/third_party/datalog/datalog.spec.in b/src/third_party/datalog/datalog.spec.in
deleted file mode 100644
index 19584e1..0000000
diff --git a/src/third_party/datalog/datalog.texi b/src/third_party/datalog/datalog.texi
deleted file mode 100644
index 86c40bc..0000000
diff --git a/src/third_party/datalog/datalogexport b/src/third_party/datalog/datalogexport
deleted file mode 100755
index 0a09de3..0000000
diff --git a/src/third_party/datalog/datalogimport b/src/third_party/datalog/datalogimport
deleted file mode 100755
index cf2a78b..0000000
diff --git a/src/third_party/datalog/depcomp b/src/third_party/datalog/depcomp
deleted file mode 100755
index debb6ff..0000000
diff --git a/src/third_party/datalog/dl_lua.h b/src/third_party/datalog/dl_lua.h
deleted file mode 100644
index b0ad0c0..0000000
diff --git a/src/third_party/datalog/even.lua b/src/third_party/datalog/even.lua
deleted file mode 100644
index 921caea..0000000
diff --git a/src/third_party/datalog/install-sh b/src/third_party/datalog/install-sh
deleted file mode 100755
index 377bb86..0000000
diff --git a/src/third_party/datalog/interp.c b/src/third_party/datalog/interp.c
deleted file mode 100644
index fb94605..0000000
diff --git a/src/third_party/datalog/laps.dl b/src/third_party/datalog/laps.dl
deleted file mode 100644
index 4a0efc7..0000000
diff --git a/src/third_party/datalog/laps.txt b/src/third_party/datalog/laps.txt
deleted file mode 100644
index d87ea5f..0000000
diff --git a/src/third_party/datalog/libtool b/src/third_party/datalog/libtool
deleted file mode 100755
index 3b00cc5..0000000
diff --git a/src/third_party/datalog/loader.c b/src/third_party/datalog/loader.c
deleted file mode 100644
index 2cf164d..0000000
diff --git a/src/third_party/datalog/long.dl b/src/third_party/datalog/long.dl
deleted file mode 100644
index 3b8dfa6..0000000
diff --git a/src/third_party/datalog/long.txt b/src/third_party/datalog/long.txt
deleted file mode 100644
index ebf5669..0000000
diff --git a/src/third_party/datalog/ltmain.sh b/src/third_party/datalog/ltmain.sh
deleted file mode 100644
index 63ae69d..0000000
diff --git a/src/third_party/datalog/lua/.deps/lapi.Po b/src/third_party/datalog/lua/.deps/lapi.Po
deleted file mode 100644
index 3d6d5be..0000000
diff --git a/src/third_party/datalog/lua/.deps/lauxlib.Po b/src/third_party/datalog/lua/.deps/lauxlib.Po
deleted file mode 100644
index eed4135..0000000
diff --git a/src/third_party/datalog/lua/.deps/lbaselib.Po b/src/third_party/datalog/lua/.deps/lbaselib.Po
deleted file mode 100644
index fbbeead..0000000
diff --git a/src/third_party/datalog/lua/.deps/lcode.Po b/src/third_party/datalog/lua/.deps/lcode.Po
deleted file mode 100644
index c0a636b..0000000
diff --git a/src/third_party/datalog/lua/.deps/ldblib.Po b/src/third_party/datalog/lua/.deps/ldblib.Po
deleted file mode 100644
index 68db24a..0000000
diff --git a/src/third_party/datalog/lua/.deps/ldebug.Po b/src/third_party/datalog/lua/.deps/ldebug.Po
deleted file mode 100644
index a67544e..0000000
diff --git a/src/third_party/datalog/lua/.deps/ldo.Po b/src/third_party/datalog/lua/.deps/ldo.Po
deleted file mode 100644
index b9155cb..0000000
diff --git a/src/third_party/datalog/lua/.deps/ldump.Po b/src/third_party/datalog/lua/.deps/ldump.Po
deleted file mode 100644
index 10257b1..0000000
diff --git a/src/third_party/datalog/lua/.deps/lfunc.Po b/src/third_party/datalog/lua/.deps/lfunc.Po
deleted file mode 100644
index 63cdcb6..0000000
diff --git a/src/third_party/datalog/lua/.deps/lgc.Po b/src/third_party/datalog/lua/.deps/lgc.Po
deleted file mode 100644
index 8af9379..0000000
diff --git a/src/third_party/datalog/lua/.deps/linit.Po b/src/third_party/datalog/lua/.deps/linit.Po
deleted file mode 100644
index 3a922c2..0000000
diff --git a/src/third_party/datalog/lua/.deps/liolib.Po b/src/third_party/datalog/lua/.deps/liolib.Po
deleted file mode 100644
index 299b285..0000000
diff --git a/src/third_party/datalog/lua/.deps/llex.Po b/src/third_party/datalog/lua/.deps/llex.Po
deleted file mode 100644
index 9ce76aa..0000000
diff --git a/src/third_party/datalog/lua/.deps/lmathlib.Po b/src/third_party/datalog/lua/.deps/lmathlib.Po
deleted file mode 100644
index 162564d..0000000
diff --git a/src/third_party/datalog/lua/.deps/lmem.Po b/src/third_party/datalog/lua/.deps/lmem.Po
deleted file mode 100644
index a9e6422..0000000
diff --git a/src/third_party/datalog/lua/.deps/loadlib.Po b/src/third_party/datalog/lua/.deps/loadlib.Po
deleted file mode 100644
index c21b57e..0000000
diff --git a/src/third_party/datalog/lua/.deps/lobject.Po b/src/third_party/datalog/lua/.deps/lobject.Po
deleted file mode 100644
index 8528c42..0000000
diff --git a/src/third_party/datalog/lua/.deps/lopcodes.Po b/src/third_party/datalog/lua/.deps/lopcodes.Po
deleted file mode 100644
index 0cef4f6..0000000
diff --git a/src/third_party/datalog/lua/.deps/loslib.Po b/src/third_party/datalog/lua/.deps/loslib.Po
deleted file mode 100644
index 4b1cf13..0000000
diff --git a/src/third_party/datalog/lua/.deps/lparser.Po b/src/third_party/datalog/lua/.deps/lparser.Po
deleted file mode 100644
index a1c22a0..0000000
diff --git a/src/third_party/datalog/lua/.deps/lstate.Po b/src/third_party/datalog/lua/.deps/lstate.Po
deleted file mode 100644
index 7586777..0000000
diff --git a/src/third_party/datalog/lua/.deps/lstring.Po b/src/third_party/datalog/lua/.deps/lstring.Po
deleted file mode 100644
index a878947..0000000
diff --git a/src/third_party/datalog/lua/.deps/lstrlib.Po b/src/third_party/datalog/lua/.deps/lstrlib.Po
deleted file mode 100644
index abaf920..0000000
diff --git a/src/third_party/datalog/lua/.deps/ltable.Po b/src/third_party/datalog/lua/.deps/ltable.Po
deleted file mode 100644
index fbc19e3..0000000
diff --git a/src/third_party/datalog/lua/.deps/ltablib.Po b/src/third_party/datalog/lua/.deps/ltablib.Po
deleted file mode 100644
index 35120b5..0000000
diff --git a/src/third_party/datalog/lua/.deps/ltm.Po b/src/third_party/datalog/lua/.deps/ltm.Po
deleted file mode 100644
index 46c2568..0000000
diff --git a/src/third_party/datalog/lua/.deps/lua.Po b/src/third_party/datalog/lua/.deps/lua.Po
deleted file mode 100644
index aa4bc92..0000000
diff --git a/src/third_party/datalog/lua/.deps/luac.Po b/src/third_party/datalog/lua/.deps/luac.Po
deleted file mode 100644
index cd95292..0000000
diff --git a/src/third_party/datalog/lua/.deps/lundump.Po b/src/third_party/datalog/lua/.deps/lundump.Po
deleted file mode 100644
index a252abb..0000000
diff --git a/src/third_party/datalog/lua/.deps/lvm.Po b/src/third_party/datalog/lua/.deps/lvm.Po
deleted file mode 100644
index 7fa7d81..0000000
diff --git a/src/third_party/datalog/lua/.deps/lzio.Po b/src/third_party/datalog/lua/.deps/lzio.Po
deleted file mode 100644
index ee9e2fd..0000000
diff --git a/src/third_party/datalog/lua/.deps/print.Po b/src/third_party/datalog/lua/.deps/print.Po
deleted file mode 100644
index 4211558..0000000
diff --git a/src/third_party/datalog/lua/COPYRIGHT b/src/third_party/datalog/lua/COPYRIGHT
deleted file mode 100644
index 84d401b..0000000
diff --git a/src/third_party/datalog/lua/Makefile.am b/src/third_party/datalog/lua/Makefile.am
deleted file mode 100644
index 037be08..0000000
diff --git a/src/third_party/datalog/lua/Makefile.in b/src/third_party/datalog/lua/Makefile.in
deleted file mode 100644
index 5c5df39..0000000
diff --git a/src/third_party/datalog/lua/README b/src/third_party/datalog/lua/README
deleted file mode 100644
index f082d40..0000000
diff --git a/src/third_party/datalog/lua/lapi.c b/src/third_party/datalog/lua/lapi.c
deleted file mode 100644
index 5d5145d..0000000
diff --git a/src/third_party/datalog/lua/lapi.h b/src/third_party/datalog/lua/lapi.h
deleted file mode 100644
index 2c3fab2..0000000
diff --git a/src/third_party/datalog/lua/lauxlib.c b/src/third_party/datalog/lua/lauxlib.c
deleted file mode 100644
index 10f14e2..0000000
diff --git a/src/third_party/datalog/lua/lauxlib.h b/src/third_party/datalog/lua/lauxlib.h
deleted file mode 100644
index 3425823..0000000
diff --git a/src/third_party/datalog/lua/lbaselib.c b/src/third_party/datalog/lua/lbaselib.c
deleted file mode 100644
index 2ab550b..0000000
diff --git a/src/third_party/datalog/lua/lcode.c b/src/third_party/datalog/lua/lcode.c
deleted file mode 100644
index 679cb9c..0000000
diff --git a/src/third_party/datalog/lua/lcode.h b/src/third_party/datalog/lua/lcode.h
deleted file mode 100644
index b941c60..0000000
diff --git a/src/third_party/datalog/lua/ldblib.c b/src/third_party/datalog/lua/ldblib.c
deleted file mode 100644
index 2027eda..0000000
diff --git a/src/third_party/datalog/lua/ldebug.c b/src/third_party/datalog/lua/ldebug.c
deleted file mode 100644
index 50ad3d3..0000000
diff --git a/src/third_party/datalog/lua/ldebug.h b/src/third_party/datalog/lua/ldebug.h
deleted file mode 100644
index ba28a97..0000000
diff --git a/src/third_party/datalog/lua/ldo.c b/src/third_party/datalog/lua/ldo.c
deleted file mode 100644
index d1bf786..0000000
diff --git a/src/third_party/datalog/lua/ldo.h b/src/third_party/datalog/lua/ldo.h
deleted file mode 100644
index 98fddac..0000000
diff --git a/src/third_party/datalog/lua/ldump.c b/src/third_party/datalog/lua/ldump.c
deleted file mode 100644
index c9d3d48..0000000
diff --git a/src/third_party/datalog/lua/lfunc.c b/src/third_party/datalog/lua/lfunc.c
deleted file mode 100644
index 813e88f..0000000
diff --git a/src/third_party/datalog/lua/lfunc.h b/src/third_party/datalog/lua/lfunc.h
deleted file mode 100644
index a68cf51..0000000
diff --git a/src/third_party/datalog/lua/lgc.c b/src/third_party/datalog/lua/lgc.c
deleted file mode 100644
index e909c79..0000000
diff --git a/src/third_party/datalog/lua/lgc.h b/src/third_party/datalog/lua/lgc.h
deleted file mode 100644
index 5a8dc60..0000000
diff --git a/src/third_party/datalog/lua/linit.c b/src/third_party/datalog/lua/linit.c
deleted file mode 100644
index c1f90df..0000000
diff --git a/src/third_party/datalog/lua/liolib.c b/src/third_party/datalog/lua/liolib.c
deleted file mode 100644
index 649f9a5..0000000
diff --git a/src/third_party/datalog/lua/llex.c b/src/third_party/datalog/lua/llex.c
deleted file mode 100644
index 88c6790..0000000
diff --git a/src/third_party/datalog/lua/llex.h b/src/third_party/datalog/lua/llex.h
deleted file mode 100644
index a9201ce..0000000
diff --git a/src/third_party/datalog/lua/llimits.h b/src/third_party/datalog/lua/llimits.h
deleted file mode 100644
index ca8dcb7..0000000
diff --git a/src/third_party/datalog/lua/lmathlib.c b/src/third_party/datalog/lua/lmathlib.c
deleted file mode 100644
index 441fbf7..0000000
diff --git a/src/third_party/datalog/lua/lmem.c b/src/third_party/datalog/lua/lmem.c
deleted file mode 100644
index ae7d8c9..0000000
diff --git a/src/third_party/datalog/lua/lmem.h b/src/third_party/datalog/lua/lmem.h
deleted file mode 100644
index 7c2dcb3..0000000
diff --git a/src/third_party/datalog/lua/loadlib.c b/src/third_party/datalog/lua/loadlib.c
deleted file mode 100644
index 6158c53..0000000
diff --git a/src/third_party/datalog/lua/lobject.c b/src/third_party/datalog/lua/lobject.c
deleted file mode 100644
index 4ff5073..0000000
diff --git a/src/third_party/datalog/lua/lobject.h b/src/third_party/datalog/lua/lobject.h
deleted file mode 100644
index f1e447e..0000000
diff --git a/src/third_party/datalog/lua/lopcodes.c b/src/third_party/datalog/lua/lopcodes.c
deleted file mode 100644
index 4cc7452..0000000
diff --git a/src/third_party/datalog/lua/lopcodes.h b/src/third_party/datalog/lua/lopcodes.h
deleted file mode 100644
index 41224d6..0000000
diff --git a/src/third_party/datalog/lua/loslib.c b/src/third_party/datalog/lua/loslib.c
deleted file mode 100644
index da06a57..0000000
diff --git a/src/third_party/datalog/lua/lparser.c b/src/third_party/datalog/lua/lparser.c
deleted file mode 100644
index dda7488..0000000
diff --git a/src/third_party/datalog/lua/lparser.h b/src/third_party/datalog/lua/lparser.h
deleted file mode 100644
index 18836af..0000000
diff --git a/src/third_party/datalog/lua/lstate.c b/src/third_party/datalog/lua/lstate.c
deleted file mode 100644
index 4313b83..0000000
diff --git a/src/third_party/datalog/lua/lstate.h b/src/third_party/datalog/lua/lstate.h
deleted file mode 100644
index 3bc575b..0000000
diff --git a/src/third_party/datalog/lua/lstring.c b/src/third_party/datalog/lua/lstring.c
deleted file mode 100644
index 4911315..0000000
diff --git a/src/third_party/datalog/lua/lstring.h b/src/third_party/datalog/lua/lstring.h
deleted file mode 100644
index 73a2ff8..0000000
diff --git a/src/third_party/datalog/lua/lstrlib.c b/src/third_party/datalog/lua/lstrlib.c
deleted file mode 100644
index 7a03489..0000000
diff --git a/src/third_party/datalog/lua/ltable.c b/src/third_party/datalog/lua/ltable.c
deleted file mode 100644
index ec84f4f..0000000
diff --git a/src/third_party/datalog/lua/ltable.h b/src/third_party/datalog/lua/ltable.h
deleted file mode 100644
index f5b9d5e..0000000
diff --git a/src/third_party/datalog/lua/ltablib.c b/src/third_party/datalog/lua/ltablib.c
deleted file mode 100644
index b6d9cb4..0000000
diff --git a/src/third_party/datalog/lua/ltm.c b/src/third_party/datalog/lua/ltm.c
deleted file mode 100644
index c27f0f6..0000000
diff --git a/src/third_party/datalog/lua/ltm.h b/src/third_party/datalog/lua/ltm.h
deleted file mode 100644
index 64343b7..0000000
diff --git a/src/third_party/datalog/lua/lua.c b/src/third_party/datalog/lua/lua.c
deleted file mode 100644
index 3a46609..0000000
diff --git a/src/third_party/datalog/lua/lua.h b/src/third_party/datalog/lua/lua.h
deleted file mode 100644
index a4b73e7..0000000
diff --git a/src/third_party/datalog/lua/luac.c b/src/third_party/datalog/lua/luac.c
deleted file mode 100644
index d070173..0000000
diff --git a/src/third_party/datalog/lua/luaconf.h b/src/third_party/datalog/lua/luaconf.h
deleted file mode 100644
index e2cb261..0000000
diff --git a/src/third_party/datalog/lua/lualib.h b/src/third_party/datalog/lua/lualib.h
deleted file mode 100644
index 469417f..0000000
diff --git a/src/third_party/datalog/lua/lundump.c b/src/third_party/datalog/lua/lundump.c
deleted file mode 100644
index 8010a45..0000000
diff --git a/src/third_party/datalog/lua/lundump.h b/src/third_party/datalog/lua/lundump.h
deleted file mode 100644
index c80189d..0000000
diff --git a/src/third_party/datalog/lua/lvm.c b/src/third_party/datalog/lua/lvm.c
deleted file mode 100644
index e0a0cd8..0000000
diff --git a/src/third_party/datalog/lua/lvm.h b/src/third_party/datalog/lua/lvm.h
deleted file mode 100644
index bfe4f56..0000000
diff --git a/src/third_party/datalog/lua/lzio.c b/src/third_party/datalog/lua/lzio.c
deleted file mode 100644
index 293edd5..0000000
diff --git a/src/third_party/datalog/lua/lzio.h b/src/third_party/datalog/lua/lzio.h
deleted file mode 100644
index 51d695d..0000000
diff --git a/src/third_party/datalog/lua/print.c b/src/third_party/datalog/lua/print.c
deleted file mode 100644
index e240cfc..0000000
diff --git a/src/third_party/datalog/main b/src/third_party/datalog/main
deleted file mode 100755
index 221d894..0000000
diff --git a/src/third_party/datalog/main.cc b/src/third_party/datalog/main.cc
deleted file mode 100644
index 12c59e1..0000000
diff --git a/src/third_party/datalog/mdate-sh b/src/third_party/datalog/mdate-sh
deleted file mode 100755
index 4614441..0000000
diff --git a/src/third_party/datalog/missing b/src/third_party/datalog/missing
deleted file mode 100755
index 9a55648..0000000
diff --git a/src/third_party/datalog/octal.dl b/src/third_party/datalog/octal.dl
deleted file mode 100644
index 041cd32..0000000
diff --git a/src/third_party/datalog/octal.txt b/src/third_party/datalog/octal.txt
deleted file mode 100644
index 1cb2e04..0000000
diff --git a/src/third_party/datalog/path.dl b/src/third_party/datalog/path.dl
deleted file mode 100644
index 30b05d7..0000000
diff --git a/src/third_party/datalog/path.txt b/src/third_party/datalog/path.txt
deleted file mode 100644
index 6230a5a..0000000
diff --git a/src/third_party/datalog/pq.dl b/src/third_party/datalog/pq.dl
deleted file mode 100644
index 3a20cdb..0000000
diff --git a/src/third_party/datalog/pq.lua b/src/third_party/datalog/pq.lua
deleted file mode 100644
index c847680..0000000
diff --git a/src/third_party/datalog/pq.txt b/src/third_party/datalog/pq.txt
deleted file mode 100644
index 7526e51..0000000
diff --git a/src/third_party/datalog/revpath.dl b/src/third_party/datalog/revpath.dl
deleted file mode 100644
index 5a44a2e..0000000
diff --git a/src/third_party/datalog/revpath.txt b/src/third_party/datalog/revpath.txt
deleted file mode 100644
index 6230a5a..0000000
diff --git a/src/third_party/datalog/says.dl b/src/third_party/datalog/says.dl
deleted file mode 100644
index eb44350..0000000
diff --git a/src/third_party/datalog/says.txt b/src/third_party/datalog/says.txt
deleted file mode 100644
index 473484a..0000000
diff --git a/src/third_party/datalog/stackless.diff b/src/third_party/datalog/stackless.diff
deleted file mode 100644
index 9371e9d..0000000
diff --git a/src/third_party/datalog/stamp-vti b/src/third_party/datalog/stamp-vti
deleted file mode 100644
index dd31729..0000000
diff --git a/src/third_party/datalog/tc.dl b/src/third_party/datalog/tc.dl
deleted file mode 100644
index ce41623..0000000
diff --git a/src/third_party/datalog/tc.txt b/src/third_party/datalog/tc.txt
deleted file mode 100644
index 0f1daff..0000000
diff --git a/src/third_party/datalog/texinfo.tex b/src/third_party/datalog/texinfo.tex
deleted file mode 100644
index a5a7b2b..0000000
diff --git a/src/third_party/datalog/true.dl b/src/third_party/datalog/true.dl
deleted file mode 100644
index ad0a6ca..0000000
diff --git a/src/third_party/datalog/true.txt b/src/third_party/datalog/true.txt
deleted file mode 100644
index 48eb7ed..0000000
diff --git a/src/third_party/datalog/try.sh b/src/third_party/datalog/try.sh
deleted file mode 100644
index 4160e30..0000000
diff --git a/src/third_party/datalog/version.texi b/src/third_party/datalog/version.texi
deleted file mode 100644
index dd31729..0000000
diff --git a/tao/auth/auth_test.go b/tao/auth/auth_test.go
index eee9fcd..0f9c0a3 100644
--- a/tao/auth/auth_test.go
+++ b/tao/auth/auth_test.go
@@ -15,6 +15,7 @@
 package auth
 
 import (
+	"encoding/base64"
 	"fmt"
 	"strings"
 	"testing"
@@ -444,3 +445,24 @@ func TestBadExtPrin(t *testing.T) {
 		}
 	}
 }
+
+var binaryEncodings = []string{
+	`DQEDa2V5BAh0ZXN0IGtleREAAQNrZXkEDHRlc3QgdGFvIGtleREA`,
+}
+
+func TestBinaryEncodings(t *testing.T) {
+	for _, b := range binaryEncodings {
+		e, err := base64.URLEncoding.DecodeString(b)
+		if err != nil {
+			t.Log(b)
+			t.Fatal("Couldn't decode a binary encoding")
+		}
+
+		f, err := UnmarshalForm(e)
+		if err != nil {
+			t.Log(b)
+			t.Fatal("Couldn't unmarshal a binary encoding")
+		}
+		t.Logf("%v\n", f)
+	}
+}
diff --git a/tao/linux_host.go b/tao/linux_host.go
index 42919cc..32f559a 100644
--- a/tao/linux_host.go
+++ b/tao/linux_host.go
@@ -158,6 +158,8 @@ func (lh *LinuxHost) Seal(child *LinuxHostChild, data []byte, policy string) ([]
 		// running on a similar LinuxHost instance. So, we don't set
 		// any policy info.
 	default:
+		// Try to parse this statement as a tao/auth policy. If it
+		// parses, then use it as the policy statement.
 		return nil, newError("policy not supported for Seal: " + policy)
 	}
 
